docs/ci_parse_aaaa_reply.3
.\"
.\" Copyright 2005 by Dominick Meglio.
.\" SPDX-License-Identifier: MIT
.\"
.TH CI_PARSE_AAAA_REPLY 3 "20 Nov 2009"
.SH NAME
ci_parse_aaaa_reply \- Parse a reply to a DNS query of type AAAA
.SH SYNOPSIS
.nf
#include <ci.h>

int ci_parse_aaaa_reply(const unsigned char *\fIabuf\fP, int \fIalen\fP,
                          struct hostent **\fIhost\fP,         
                          struct ci_addr6ttl *\fIaddrttls\fP, int *\fInaddrttls\fP);
.fi
.SH DESCRIPTION
The
.B ci_parse_aaaa_reply
function parses the response to a query of type AAAA into a
.BR "struct hostent"
and/or an array of
.BR "struct ci_addr6ttl" . 
The parameters
.I abuf
and
.I alen
give the contents of the response.  The result is stored in allocated
memory and a pointer to it stored into the variable pointed to by
.IR host ,
if host is nonnull.
It is the caller's responsibility to free the resulting host structure
using
.BR ci_free_hostent (3)
when it is no longer needed.
.PP
If
.IR addrttls
and
.IR naddrttls
are both nonnull,
then up to *naddrttls
.BR "struct ci_addr6ttl"
records are stored in the array pointed to by addrttls,
and then *naddrttls is set to the number of records so stored.
Note that the memory for these records is supplied by the caller.
.SH RETURN VALUES
.B ci_parse_aaaa_reply
can return any of the following values:
.TP 15
.B CI_SUCCESS
The response was successfully parsed.
.TP 15
.B CI_EBADRESP
The response was malformatted.
.TP 15
.B CI_ENODATA
The response did not contain an answer to the query.
.TP 15
.B CI_ENOMEM
Memory was exhausted.
.SH SEE ALSO
.BR ci_gethostbyname (3),
.BR ci_free_hostent (3)


RELEASE-NOTES.md
## c-ci version 1.34.2 - October 15 2024

This release contains a fix for downstream packages detecting the c-ci
version based on the contents of the header file rather than the
distributed pkgconf or cmake files.

## c-ci version 1.34.1 - October 9 2024

This release fixes a packaging issue.


## c-ci version 1.34.0 - October 9 2024

This is a feature and bugfix release.

Features:
* adig: read arguments from adigrc.
  [PR #856](https://github.com/c-ci/c-ci/pull/856)
* Add new pending write callback optimization via `ci_set_pending_write_cb`.
  [PR #857](https://github.com/c-ci/c-ci/pull/857)
* New function `ci_process_fds()`.
  [PR #875](https://github.com/c-ci/c-ci/pull/875)
* Failed servers should be probed rather than redirecting queries which could
  cause unexpected latency.
  [PR #877](https://github.com/c-ci/c-ci/pull/877)
* adig: rework command line arguments to mimic dig from bind.
  [PR #890](https://github.com/c-ci/c-ci/pull/890)
* Add new method for overriding network functions
  `ci_set_socket_function_ex()` to properly support all new functionality.
  [PR #894](https://github.com/c-ci/c-ci/pull/894)
* Fix regression with custom socket callbacks due to DNS cookie support.
  [PR #895](https://github.com/c-ci/c-ci/pull/895)
* ci_socket: set IP_BIND_ADDRESS_NO_PORT on ci_set_local_ip* tcp sockets
  [PR #887](https://github.com/c-ci/c-ci/pull/887)
* URI parser/writer for ci_set_servers_csv()/ci_get_servers_csv().
  [PR #882](https://github.com/c-ci/c-ci/pull/882)

Changes:
* Connection handling modularization.
  [PR #857](https://github.com/c-ci/c-ci/pull/857),
  [PR #876](https://github.com/c-ci/c-ci/pull/876)
* Expose library/utility functions to tools.
  [PR #860](https://github.com/c-ci/c-ci/pull/860)
* Remove `ci__` prefix, just use `ci_` for internal functions.
  [PR #872](https://github.com/c-ci/c-ci/pull/872)


Bugfixes:
* fix: potential WIN32_LEAN_AND_MEAN redefinition.
  [PR #869](https://github.com/c-ci/c-ci/pull/869)
* Fix googletest v1.15 compatibility.
  [PR #874](https://github.com/c-ci/c-ci/pull/874)
* Fix pkgconfig thread dependencies.
  [PR #884](https://github.com/c-ci/c-ci/pull/884)


Thanks go to these friendly people for their efforts and contributions for this
release:

* Brad House (@bradh352)
* Cristian Rodríguez (@crrodriguez)
* Georg (@tacerus)
* @lifenjoiner
* Shelley Vohr (@codebytere)
* 前进，前进，进 (@leleliu008)



src/lib/legacy/ci_parse_aaaa_reply.c
/* MIT License
 *
 * Copyright (c) 1998 Massachusetts Institute of Technology
 * Copyright (c) 2005 Dominick Meglio
 * Copyright (c) 2019 Andrew Selivanov
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * SPDX-License-Identifier: MIT
 */

#include "ci_private.h"

#ifdef HAVE_NETINET_IN_H
#  include <netinet/in.h>
#endif
#ifdef HAVE_NETDB_H
#  include <netdb.h>
#endif
#ifdef HAVE_ARPA_INET_H
#  include <arpa/inet.h>
#endif

#ifdef HAVE_STRINGS_H
#  include <strings.h>
#endif

#ifdef HAVE_LIMITS_H
#  include <limits.h>
#endif

#include "ci_inet_net_pton.h"

int ci_parse_aaaa_reply(const unsigned char *abuf, int alen,
                          struct hostent **host, struct ci_addr6ttl *addrttls,
                          int *naddrttls)
{
  struct ci_addrinfo ai;
  char                *question_hostname = NULL;
  ci_status_t        status;
  size_t               req_naddrttls = 0;
  ci_dns_record_t   *dnsrec        = NULL;

  if (alen < 0) {
    return CI_EBADRESP;
  }

  if (naddrttls) {
    req_naddrttls = (size_t)*naddrttls;
    *naddrttls    = 0;
  }

  memset(&ai, 0, sizeof(ai));

  status = ci_dns_parse(abuf, (size_t)alen, 0, &dnsrec);
  if (status != CI_SUCCESS) {
    goto fail;
  }

  status = ci_parse_into_addrinfo(dnsrec, 0, 0, &ai);
  if (status != CI_SUCCESS && status != CI_ENODATA) {
    goto fail;
  }

  if (host != NULL) {
    status = ci_addrinfo2hostent(&ai, AF_INET6, host);
    if (status != CI_SUCCESS && status != CI_ENODATA) {
      goto fail; /* LCOV_EXCL_LINE: DefensiveCoding */
    }
  }

  if (addrttls != NULL && req_naddrttls) {
    size_t temp_naddrttls = 0;
    ci_addrinfo2addrttl(&ai, AF_INET6, req_naddrttls, NULL, addrttls,
                          &temp_naddrttls);
    *naddrttls = (int)temp_naddrttls;
  }

fail:
  ci_freeaddrinfo_cnames(ai.cnames);
  ci_freeaddrinfo_nodes(ai.nodes);
  ci_free(question_hostname);
  ci_free(ai.name);
  ci_dns_record_destroy(dnsrec);

  if (status == CI_EBADNAME) {
    status = CI_EBADRESP;
  }

  return (int)status;
}
